### PR TITLE
KPP standalone interface from Obin Sturm

### PR DESCRIPTION
**Pull request by lizziel**
### Name and Institution (Required)

Name: Lizzie Lundgren
Institution: Harvard University

### Describe the update

This draft PR contains updates for KPP standalone interface from Obin Sturm (GMAO) contained in https://github.com/GEOS-ESM/geos-chem/pull/9. That original PR is built off of branch geos/develop in the GEOS-Chem fork in GMAO's GitHub. That version is not mergable into the standard model so this PR will be used instead.

Tagging @obin1, @yantosca 

### Expected changes

I have pushed several updates to this PR:

1.  `kpp_standalone_interface.yml` configuration file:

    - Comments have been updated
    - This is now always copied to any fullchem run directory (GCClassic only).
    - The following YAML tags have been added.  You now must manually activate the KPP standalone interface.  Also, you can set the interval during which model state will be archived.  By default this is set to the interval 1900..2100, but you can manually restrict this to e.g. the last timestep of a run.
    ```yaml
    settings:
      activate: true                       # Main on/off switch
      start_output_at: [19000101, 000000]  # Save model state for KPP standalone
      stop_output_at:  [21000101, 000000]  # ... if between these 2 datetimes
    ```

2. `GeosCore/kpp_standalone_interface.yml` has been renamed to `GeosCore/kppsa_interface_mod.F90`.  All derived types, variables, and public-facing routines have been renamed to use the prefix `KppSa_`, for consistency.   Derived-type 
 object `Kpp_Standalone_YAML` has new been renamed to `KppSa_State` for brevity.

3. We have added a new derived-type object `KppSa_ActiveCell` to hold the `ActiveCell` logical flag and `Active_Cell_Name` fields.  This is declared with `!$OMP THREADPRIVATE` because it is updated within the main parallel loop within the `DO_FULLCHEM` routine.

4. In the `KppSa_Write_Samples` routine (formerly `Write_Samples`), we have placed the I/O code within an `!$OMP CRITICAL` block.  This will prevent other threads from trying to write to the file, which can cause incomplete and/or corrupted files.

5. We have added a new function `KppSa_Check_Time` to determine if we are within the time window for archiving model state specified in the YAML file.  This updates a new flag `KppSa_State%SkipWriteAtThisTime`, which is used to determine if we need to skip writing output on this timestep.

6. We now print the following output at initialization:
```console
===============================================================================
KPP STANDALONE INTERFACE

Model state will be archived at these sites:

LosAngeles               ( -118.2430,   34.0522 )
McMurdo                  (  166.6698,  -77.8455 )
Paris                    (    2.3522,   48.8566 )
Beijing                  (  116.4074,   39.9042 )
Kinshasa                 (   15.3105,   -4.3033 )
Kennaook                 (  144.6833,  -40.6833 )
Graciosa                 (  -28.0069,   39.0525 )
Utqiagvik                ( -156.7886,   71.2906 )
Ozarks                   (  -91.2590,   37.5020 )
Amazon                   (  -62.2159,   -3.4653 )
Congo                    (   12.5484,   -5.9175 )
Borneo                   (  114.0000,    0.0000 )
IndianOcean              (   87.2000,   23.0000 )
AtlanticOcean            (  -41.5748,   34.7079 )
PacificOcean             ( -121.9645,    0.0000 )
ElDjouf                  (   -6.6661,   21.5008 )

For GEOS-Chem vertical levels:
   1   2  10  23  35  48  56
Starting at 20190701 004000
Ending at   20190701 010000
===============================================================================
```

### Reference(s)
None

### Related Github Issue
- See https://github.com/GEOS-ESM/geos-chem/pull/9

### CODE REVIEW

**Comment by obin1** [comment]
Hi Lizzie, thanks for getting this started! I'm still working on redoing a few c360 runs with the new compute resources on Discover before we can write up the manuscript about this tool and species-level tolerances. I will keep you all posted when that's out as a preprint, and thanks again for moving this PR upstream

File: run/shared/kpp_standalone_interface.yml [hunk 1]
```diff
@@ -0,0 +1,77 @@
+active_cells:
```
Comment (yantosca) [inline run/shared/kpp_standalone_interface.yml:1]: @obin1, I just have a couple of thoughts.

1. Right now this file isn't automatically copied when we create a run directory.  That's an easy fix though, we'd just have to edit `run/GCClassic/createRunDir.sh` accordingly.
2. I think it'd be nice to have a master on/off switch either in this file or in `geoschem_config.yml`.  If we copy this file to the run directory, right now there is no way to disable saving this output (which most users probably won't want to do.)   
Interested to hear your thoughts.

**Review by yantosca — COMMENTED** [review]
@obin1: Also not related to this PR, but if you have any python scripts to read & plot output from the box model we shoudl think about bringing those into GCPy.

**Comment by yantosca** [comment]
@obin1 @lizziel:  Maybe this could be for the KPP interface 2.0, but it'd be really nice to specify the time at which you would like to write out the atmospheric state.  You might only really need to do this at the end of a run rather than on every timestep, if the goal is to obtain well-spun-up initial conditions for the box model.

**Comment by yantosca** [comment]
@obin1 @lizziel: Also I've noticed we get a bunch of fort.* files.  I'll try to track this down.  This usually occurs when you open a file but don't specify the filename.
```console
test_env) [holy8a24301 test_kppbox]$ ls
archiveRun.sh*     fort.12     gcclassic_timers.json           HISTORY.rc
build/             fort.13     GC.log                          kpp_standalone_interface.yml
build_info/        fort.14     geoschem.benchmark.run*         metrics.py*
cleanRunDir.sh*    fort.15     geoschem_config.yml             OutputDir/
CodeDir@           fort.16     getRunInfo*                     README.md
CreateRunDirLogs/  fort.17     HEMCO_Config.rc                 Restarts/
download_data.py*  fort.18     HEMCO_Config.rc.gmao_metfields  runScriptSamples@
download_data.yml  gcclassic*  HEMCO_Diagn.rc                  species_database.yml
```

**Comment by yantosca** [comment]
@obin1: @lizziel I'm finding that sometimes a file gets written without the site name.  I added some debug output to the `Write_Samples` routine:
```F90
      !========================================================================
      ! Write the file
      !========================================================================
      ! Find a free file LUN
      IU_FILE = findFreeLUN()
      write(level_string,'(I0)') L
      write(YYYYMMDD_hhmmz,'(I0.4,I0.2,I0.2,a,I0.2,I0.2)' ) &
            Get_Year(), Get_Month(), Get_Day(),'_', Get_Hour(), Get_Minute()

      filename = trim(KPP_Standalone_YAML%Output_Directory)//'/'               &
                      //trim(CELL_NAME_AUX)//trim(KPP_Standalone_YAML%ACTIVE_CELL_NAME) &
                      //'_L'//trim(level_string)//'_' //trim(YYYYMMDD_hhmmz)//'.txt'
      print*, '%%%', trim(filename)                  
```
and when I run for an hour I get this output
```console
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/McMurdo_L1_20190701_0000.txt
 %%%./OutputDir/Kennaook_L1_20190701_0000.txt
 %%%./OutputDir/Kennaook_L1_20190701_0000.txt
 %%%./OutputDir/Amazon_L1_20190701_0000.txt
 %%%./OutputDir/Borneo_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/Congo_L1_20190701_0000.txt
 %%%./OutputDir/Congo_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/IndianOcean_L1_20190701_0000.txt
 %%%./OutputDir/IndianOcean_L1_20190701_0000.txt
 %%%./OutputDir/IndianOcean_L1_20190701_0000.txt
 %%%./OutputDir/IndianOcean_L1_20190701_0000.txt
 %%%./OutputDir/IndianOcean_L1_20190701_0000.txt
 %%%./OutputDir/IndianOcean_L1_20190701_0000.txt
 %%%./OutputDir/IndianOcean_L1_20190701_0000.txt
 %%%./OutputDir/ElDjouf_L1_20190701_0000.txt
 %%%./OutputDir/ElDjouf_L1_20190701_0000.txt
 %%%./OutputDir/ElDjouf_L1_20190701_0000.txt
 %%%./OutputDir/ElDjouf_L1_20190701_0000.txt
 %%%./OutputDir/ElDjouf_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/AtlanticOcean_L1_20190701_0000.txt
 %%%./OutputDir/AtlanticOcean_L1_20190701_0000.txt
 %%%./OutputDir/AtlanticOcean_L1_20190701_0000.txt
 %%%./OutputDir/AtlanticOcean_L1_20190701_0000.txt
 %%%./OutputDir/AtlanticOcean_L1_20190701_0000.txt
 %%%./OutputDir/AtlanticOcean_L1_20190701_0000.txt
 %%%./OutputDir/AtlanticOcean_L1_20190701_0000.txt
 %%%./OutputDir/AtlanticOcean_L1_20190701_0000.txt
 %%%./OutputDir/PacificOcean_L1_20190701_0000.txt
 %%%./OutputDir/PacificOcean_L1_20190701_0000.txt
 %%%./OutputDir/PacificOcean_L1_20190701_0000.txt
 %%%./OutputDir/Graciosa_L1_20190701_0000.txt
 %%%./OutputDir/Graciosa_L1_20190701_0000.txt
 %%%./OutputDir/Graciosa_L1_20190701_0000.txt
 %%%./OutputDir/Graciosa_L1_20190701_0000.txt
 %%%./OutputDir/Graciosa_L1_20190701_0000.txt
 %%%./OutputDir/Graciosa_L1_20190701_0000.txt
 %%%./OutputDir/Graciosa_L1_20190701_0000.txt
 %%%./OutputDir/Graciosa_L1_20190701_0000.txt
 %%%./OutputDir/Graciosa_L1_20190701_0000.txt
 %%%./OutputDir/Beijing_L1_20190701_0000.txt
 %%%./OutputDir/Beijing_L1_20190701_0000.txt
 %%%./OutputDir/Beijing_L1_20190701_0000.txt
 %%%./OutputDir/Beijing_L1_20190701_0000.txt
 %%%./OutputDir/Beijing_L1_20190701_0000.txt
 %%%./OutputDir/Beijing_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Ozarks_L1_20190701_0000.txt
 %%%./OutputDir/Paris_L1_20190701_0000.txt
 %%%./OutputDir/_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
 %%%./OutputDir/Utqiagvik_L1_20190701_0000.txt
```
Not sure why some cell names are being dropped.  It seems like this might be a race condition.  If this is called from a parallel loop them maybe we are missing some private variables.  I'll dig further.

**Comment by yantosca** [comment]
@obin1 @lizziel: See the updates in my bugfix branch below.  This fixes the parallelization issue listed above.

- https://github.com/geoschem/geos-chem/tree/bugfix/kpp_box_fixes

File: run/shared/kpp_standalone_interface.yml (hunk 1 above)
Comment (obin1) [reply run/shared/kpp_standalone_interface.yml:30]: Hi @yantosca, sorry for missing this earlier! 
Regarding point 1, I've never created a GEOS-Chem run directory, we've just been copying prior GEOS-CF run directories. Automatically copying sounds like the way to go.
For point 2: If I understand some subsequent commits, you've made that on/off switch. That is better than what I've been doing, which is just renaming the file when I didn't want to print output.

**Comment by yantosca** [comment]
@lizziel, @obin1: I have confirmed that this PR works in GCHP.  I was able to restrict output to the last timestep of a 1-hr GCHP run and got a bunch of files timestamped at 00:40z.

File: GeosCore/fullchem_mod.F90 [hunk 2]
```diff
@@ -583,9 +600,8 @@ SUBROUTINE Do_FullChem( Input_Opt,  State_Chm, State_Diag,                 &
 
        ! Check if the current grid cell in this loop should have its
        ! full chemical state printed (concentrations, rates, constants)
-       ! for use with the KPP Standalone
-       ! (psturm, 03/22/24)
-       CALL Check_ActiveCell( I, J, L, State_Grid )
+       ! for use with the KPP Standalone (psturm, 03/22/24)
+       CALL KppSa_Check_ActiveCell( I, J, L )
```
Comment (obin1) [inline GeosCore/fullchem_mod.F90:604]: Good call, State_Grid was only used for debugging. If users want to see how far their sampled grid cell midpoints are from the assigned locations, they can compare output files to the yaml file coordinates.

**Comment by yantosca** [comment]
@obin1 @lizziel: FYI I have also modified this PR to save out the `ICNTRL` and `RCNTRL` solver options to the output file.  This is necessary in case we want to use different solvers.

Now that we have per-species ATOL settings in GEOS-Chem 14.5.0, I will add another fix to save that out.  This will be needed to set the ATOL for family/dummy species to a high value so as to keep them out of the error norm computations.

**Review by yantosca — APPROVED** [review]

**Comment by yantosca** [comment]
All GCClassic integration tests passed:
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

CodeDir   : 0ecd8fa GEOS-Chem update: Merge PR #2507 (Fix CEDS TMB emissions)
GEOS-Chem : a86303771 Merge PR #2482 (KPP standalone interface from Obin Sturm)
HEMCO     : deaa192 HEMCO 3.10.0 release
Cloud-J   : f8a2b7f Update version number for 8.0.1 release
HETP      : 2a99b24 Merge pull request #2 from geoschem/bugfix/initialize_local_variables

Using 24 OpenMP threads
Number of execution tests: 30

Submitted as SLURM job: 58916604
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
Also all tests were zero-diff w/r/t 14.6.0-alpha.0 except for
- TOMAS15 (parallelization error?)
- APM (parallelization error?)

**Comment by yantosca** [comment]
All GCHP integration tests passed:
```console
==============================================================================
GCHP: Execution Test Results

CodeDir       : 15de6b3 GEOS-Chem update: Merge PR #2507 (Fix CEDS TMB emissions)
MAPL          : 9ad63ae Merge PR #37 containing update to vertically flip imports with dimensionless pressure proxy lev coordinates
GMAO_Shared   : 4ddb3ec Merge pull request #2 from geoschem/feature/mapl-upgrade
ESMA_cmake    : ad5deba Added ecbuild as a submodule of ESMA_cmake
gFTL-shared   : 4b82492 Merge branch 'upstream_v1.5.0' into feature/v1.5.0
FMS           : 259759d Merge pull request #3 from geoschem/feature/update_gmao_libs
FVdycoreCubed : af42462 Merge PR #8 (Add PLEadv diagnostic for offline advection in GCHP)
geos-chem     : a86303771 Merge PR #2482 (KPP standalone interface from Obin Sturm)
HEMCO         : deaa192 HEMCO 3.10.0 release
yaFyaml       : 19afe50 Merge branch 'upstream_v1.0.4' into feature/v1.0.4
pFlogger      : 2c4b724 Merge branch 'upstream_v1.9.1' into feature/v1.9.1
Cloud-J       : f8a2b7f Update version number for 8.0.1 release
HETP          : 2a99b24 Merge pull request #2 from geoschem/bugfix/initialize_local_variables

Number of execution tests: 12

Submitted as SLURM job: 58926494
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
All tests were zero-diff w/r/t 14.6.0-alpha.1
